internal/controller/customreplicaset_controller.go

```diff
@@ -74,13 +80,130 @@ func (r *CustomReplicaSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{Requeue: true}, err
 	}
 
 	return ctrl.Result{Requeue: false}, err
 }
 
+// Create a new ControllerRevision
+func (r *CustomReplicaSetReconciler) createControllerRevision(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, latestRevision *v1.ControllerRevision, jsonSpec []byte) error {
+	var revisionNumber int64
+	if latestRevision == nil {
+		revisionNumber = 0
+	} else {
+		revisionNumber = latestRevision.Revision + 1
+	}
+
+	t := time.Now()
+	timestamp := fmt.Sprintf("%d%d%d%d", t.Hour(), t.Minute(), t.Second(), t.Nanosecond())
+
+	newRevision := v1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-controller-revision-" + timestamp,
+			Namespace: cr.Namespace,
+			Labels:    cr.Labels,
+		},
+		Data:     runtime.RawExtension{Raw: jsonSpec},
+		Revision: revisionNumber,
+	}
+	if err := controllerutil.SetControllerReference(cr, &newRevision, r.Scheme); err != nil {
+		return err
+	}
+	if err := r.Create(ctx, &newRevision); err != nil {
+		return err
+	}
+
+	// Update 'revisionNumber' label of the CustomReplicaSet
+	if cr.Labels == nil {
+		cr.Labels = map[string]string{}
+	}
+	cr.Labels["revisionNumber"] = strconv.FormatInt(newRevision.Revision, 10)
+	if err := r.Update(ctx, cr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *CustomReplicaSetReconciler) manageControllerRevisionHistory(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, latestRevision *v1.ControllerRevision) error {
+	if latestRevision == nil {
+		jsonSpec, err := convertSpecToJson(cr)
+		if err != nil {
+			return err
+		}
+
+		if err := r.createControllerRevision(ctx, cr, latestRevision, jsonSpec); err != nil {
+			fmt.Println("Failed to create first controller revision")
+			return err
+		}
+		return nil
+	}
+	var latestSpec customreplicasetv1.CustomReplicaSetSpec
+
+	if err := json.Unmarshal(latestRevision.Data.Raw, &latestSpec); err != nil {
+		fmt.Println("Failed to unmarshal latest controller spec to json")
+		return err
+	}
+
+	// Compare the PodTemplates
+	if !reflect.DeepEqual(cr.Spec.Template, latestSpec.Template) {
+		jsonSpec, err := convertSpecToJson(cr)
+		if err != nil {
+			return err
+		}
+
+		if err := r.createControllerRevision(ctx, cr, latestRevision, jsonSpec); err != nil {
+			fmt.Println("Failed to create new controller revision")
+			return err
+		}
+	}
+	return nil
+}
+
+func convertSpecToJson(cr *customreplicasetv1.CustomReplicaSet) ([]byte, error) {
+	jsonSpec, err := json.Marshal(cr.Spec)
+	if err != nil {
+		fmt.Println("Failed to convert CustomReplicaSet spec to json")
+		return nil, err
+	}
+	return jsonSpec, nil
+}
+
+func (r *CustomReplicaSetReconciler) getLatestControllerRevision(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet) (*v1.ControllerRevision, error) {
+	controllerRevisions := &v1.ControllerRevisionList{}
+	if err := r.List(ctx, controllerRevisions, client.InNamespace(cr.Namespace)); err != nil {
```

Comment:
You should only list the controller revision that belongs to the cr.

Reply:
Yes added an owner label, such that only controller revisions belonging to the CR are listed. 

`r.List(ctx, controllerRevisions, client.InNamespace(cr.Namespace), client.MatchingLabels{"owner": cr.Name})`